bftengine/include/bftengine/DbCheckpointMetadata.hpp

```diff
@@ -0,0 +1,62 @@
+// Concord
+//
+// Copyright (c) 2021 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the
+// "License").  You may not use this product except in compliance with the
+// Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#pragma once
+
+#include "Serializable.h"
+
+#include <vector>
+#include <chrono>
+
+namespace bftEngine::impl {
+constexpr uint32_t MAX_ALLOWED_CHECKPOINTS{100};
+using CheckpointId = uint64_t;
+using TimeDuration = std::chrono::duration<long>;
+
+struct DbCheckpointMetadata : public concord::serialize::SerializableFactory<DbCheckpointMetadata> {
```

Comment:
Do we really need to persist this to a mdt storage?
Can't we derive the checkpoints from the checkpoints directory of database itself?

Reply:
There can be a case, when replica crashes while creating db checkpoint. If we find, a checkpoint directory on filesystem but no matching entry in dbCheckpointMetadata, then its probably an in-consistent db checkpoint, so we remove it. Also, we store last checkpoint creation time, for creation of checkpoint after some configured duration